src/contrib/tevent.rst

```diff
@@ -0,0 +1,655 @@
+Tutorial`_ and `API Reference`_ for more information on registering event
+handlers.
+
+.. code-tabs::
+
+    .. code-tab:: c
+        :label: File descriptor example
+        :class: short
+
+        /**
+         * Read from a remote connection whenever some data is ready.
+         */
+
+        #include <arpa/inet.h>
+        #include <errno.h>
+        #include <netinet/in.h>
+        #include <stdio.h>
+        #include <string.h>
+        #include <sys/socket.h>
+        #include <tevent.h>
+        #include <unistd.h>
+
+        int open_socket(int *_fd)
+        {
+            struct sockaddr_in addr;
+            int ret;
+            int fd;
+
+            fd = socket(AF_INET, SOCK_STREAM, 0);
+            if (fd == -1) {
+                fprintf(stderr, "Unable to create socket\n");
+                return EIO;
+            }
+
+            addr.sin_family = AF_INET;
+            addr.sin_port = htons(3333);
+            addr.sin_addr.s_addr = inet_addr("127.0.0.1");
+
+            ret = connect(fd, (struct sockaddr *)&addr, sizeof(struct sockaddr_in));
+            if (ret != 0) {
+                ret = errno;
+                fprintf(stderr, "Unable to connect to localhost:3333 [%d]: %s\n",
+                        ret, strerror(ret));
+                return ret;
+            }
+
+            *_fd = fd;
+
+            return 0;
+        }
+
+        static void fd_readable_handler(struct tevent_context *ev,
+                                        struct tevent_fd *fde,
+                                        uint16_t flags,
+                                        void *pvt)
+        {
+            int fd = *(int*)pvt;
+            char buf[255] = {0};
+            int ret;
+
+            puts("File descriptor is readable!");
+            ret = read(fd, &buf, 254);
+            printf("ret = %d, errno = %d\n", ret, errno);
+            printf("Data read: %s\n", buf);
+        }
+
+        int main()
+        {
+            struct tevent_context *ev;
+            struct tevent_fd *fde;
+            int ret;
+            int fd;
+
+            /* Initialize tevent context. */
+            ev = tevent_context_init(NULL);
+            if (ev == NULL) {
+                puts("Unable to create tevent context.");
+                return 1;
+            }
+
+            /* Open socket. */
+            ret = open_socket(&fd);
+            if (ret != 0) {
+                puts("Unable to open socket.");
+                return 1;
+            }
+
+            /* Register event handler - executed when the fd is readable */
+            fde = tevent_add_fd(ev, NULL, fd, TEVENT_FD_READ, fd_readable_handler, &fd);
+            if (fde == NULL) {
+                puts("Unable to create fde.");
+                return 1;
+            }
+
+            /* Enter the event loop. */
+            while (true) {
+                tevent_loop_wait(ev);
+            }
+        }
+
+        /* You can test it using the ncat utility. */
```

Comment:
If this is possible to split code snippet and bash snippet under the same tab?

Reply:
Done.